### PR TITLE
Process owner id metadata if present

### PR DESCRIPTION
**Pull request by respinos**
Invoke `ProcessOwnerIds` when the MdpItem is built, and various methods to get seq/ownerid.

This requires updating the cache key.

### CODE REVIEW

File: MdpItem.pm [hunk 1]
```diff
@@ -1493,6 +1493,80 @@ sub GetItemCover {
...
+                my $tree = $parser->parse_string($$metsXmlRef);
+                my $root = $tree->getDocumentElement();
+
+                foreach my $file ( $root->findnodes(qq{//METS:file[\@OWNERID]}) ) {
+                    my $ownerid = $file->getAttribute('OWNERID');
+                    next unless ( $ownerid );
+                    my @ownerids = split(/ /, $ownerid);
+                    my $fileid = $file->getAttribute('ID');
+                    my $seq = $root->findvalue(qq{//METS:structMap[\@TYPE="physical"]//METS:div[METS:fptr[\@FILEID="$fileid"]]/\@ORDER});
+                    $$seq_map{$seq} = $ownerids[0];
```
Comment (respinos) [inline MdpItem.pm]: @aelkiss would the most recent ownerid be listed first or last?

File: MdpItem.pm [hunk 2]
```diff
@@ -188,7 +188,7 @@ sub __handle_mdpitem_cache_setup {
         $cache_max_age = $config->get('mdpitem_max_age') || 0;
     }
 
-    my $cache_key = qq{mdpitem};
+    my $cache_key = qq{mdpitem.b};
```
Comment (aelkiss) [inline MdpItem.pm:191]: Typo? "mdpitem.b" doesn't show up anywhere else as far as I can tell

Comment (respinos) [reply MdpItem.pm:191]: Nope; it's just the new cache_key name --- that value gets passed back to `GetMdpItem` and used there.
(The usual ways to invalidate the cache aren't available, alas.)

File: MdpItem.pm (hunk 1 above)
Comment (aelkiss) [reply MdpItem.pm]: There will only be one `OWNERID` on `//METS:file`, which represents the `OWNERID` of the page scan actually in use.

In looking at an example (chi.082891291) I see:

```xml
<METS:file ID="IMG00000001" MIMETYPE="image/jp2" SEQ="00000001" CREATED="2014-10-06T00:00:00" SIZE="409054" CHECKSUM="93a5b01fb43835d0d1a283660108cbe0" CHECKSUMTYPE="MD5" OWNERID="13510798903268757-5" ADMID="IMAGE_METHOD0 ADD_PAGE_DATA_IMG00000001">
        <METS:FLocat xlink:href="00000001.jp2" LOCTYPE="OTHER" OTHERLOCTYPE="SYSTEM"/>
      </METS:file>
```
Other candidate `OWNERID`s that are believed to be the same page will be listed in the referenced `techMD`:

```xml
    <METS:techMD ID="ADD_PAGE_DATA_IMG00000001">
      <METS:mdWrap LABEL="candidates" MDTYPE="OTHER" OTHERMDTYPE="Google">
        <METS:xmlData>
          <gbs:pageNo>PP1</gbs:pageNo>
          <gbs:scanIDwrap>
            <gbs:ownerID>13510798903268757-5</gbs:ownerID>
          </gbs:scanIDwrap>
        </METS:xmlData>
      </METS:mdWrap>
    </METS:techMD>
```

I need to dig up an example, but I believe there will just be a sequence of `gbs:ownerID` elements that lists all candidates.

So, I think the way to get other OWNERIDs that should map to this sequence as well is to look for an element referenced by the `ADMID` that contains an `mdWrap[@LABEL="candidates"]` (and sadly I don't think there's a way to tokenize those IDREF values directly via XPath 1.0, which is all libxml2 supports) and then get all its `gbs:ownerID` descendants -- or equivalently to find the page a specific OWNERID should reference, find the `gbs:candidate` element that contains it, then find the file that references the enclosing techMD...

**Review by aelkiss — CHANGES_REQUESTED** [review]
Doesn't currently fetch all potential OWNERIDs -- see comment

File: MdpItem.pm (hunk 1 above)
Comment (respinos) [reply MdpItem.pm]: @aelkiss changed the code to collect all the `gbs:ownerID` values that belong to a `techMD` element before processing the `file` elements.

File: MdpItem.pm [hunk 3]
```diff
@@ -1493,6 +1493,106 @@ sub GetItemCover {
...
+                return unless ( scalar @techmd_els );
+
+                my $techmd_ownerid_map = {};
+                foreach my $techmd_el ( @techmd_els ) {
+                    my $id = $techmd_el->getAttribute('ID');
+                    my @tmp = ();
+                    foreach my $ownerid_el ( $techmd_el->findnodes(qq{.//gbs:ownerID}) ) {
+                        push @tmp, $ownerid_el->textContent;
+                    }
+                    $$techmd_ownerid_map{$id} = [ sort @tmp ];
```
Comment (aelkiss) [inline MdpItem.pm:1532]: $$techmd_ownerid_map will look like e.g. 

```perl
{
   ADD_PAGE_DATA_IMG00000001 => [ 'SOME_OWNERID',  'ANOTHER_OWNERID', ... ], 
   ADD_PAGE_DATA_IMG00000002 => ['MORE_OWNERIDS', 'YET_ANOTHER_OWNERID', ...]
}
```

etc

File: MdpItem.pm [hunk 4]
```diff
@@ -1493,6 +1493,106 @@ sub GetItemCover {
...
+
+                foreach my $file ( $root->findnodes(qq{//METS:file[\@OWNERID]}) ) {
+                    my $ownerid = $file->getAttribute('OWNERID');
+                    next unless ( $ownerid );
+
+                    my $fileid = $file->getAttribute('ID');
+
+                    my @admids = split(/ /, $file->getAttribute('ADMID'));
+                    my $admid;
+                    foreach ( @admids ) {
```
Comment (aelkiss) [inline MdpItem.pm:1543]: Perhaps:

```perl
use List::Util qw(first);

my $admid = first { defined $$techid_ownerid_map{$_} } @admids;
```

File: MdpItem.pm [hunk 5]
```diff
@@ -1493,6 +1493,106 @@ sub GetItemCover {
...
+                    }
+                    next unless ( $admid );
+
+                    my $seq = $root->findvalue(qq{//METS:structMap[\@TYPE="physical"]//METS:div[METS:fptr[\@FILEID="$fileid"]]/\@ORDER});
+                    foreach my $ownerid ( @{ $$techmd_ownerid_map{$admid} } ) {
+                        $$ownerid_map{$ownerid} = $seq;
+                    }
+
+                    # map the sequence to the ownerid in use
+                    $$seq_map{$seq} = $ownerid;
```
Comment (aelkiss) [inline MdpItem.pm:1557]: After looking this through and checking against an actual METS file I'm pretty sure this is correct.

**Review by aelkiss — APPROVED** [review]

**Pull request merged**